File: Problems/0563_Binary_Tree_Tilt/Project_Scala/Operate_TreeNode.scala
object Operate_TreeNode {
    def set_TreeNode(flds:Array[String]):TreeNode = {
        return set_TreeNode(flds, 0, 0)
    }

    def set_TreeNode(flds:Array[String], depth:Int, pos:Int):TreeNode = {
        if (flds.size <= 0)
            return null

        var cur_pos:Int = 0
        for (i <- 0 until depth)
            cur_pos = cur_pos + Math.pow(2, i).toInt
        
        if (cur_pos + pos > flds.size - 1)
            return null
        
        if (flds(cur_pos + pos) == "null")
            return null

        var node:TreeNode = new TreeNode(flds(cur_pos + pos).toInt)
        node.left = set_TreeNode(flds, depth + 1, 2*pos)
        node.right = set_TreeNode(flds, depth + 1, 2*pos + 1)

        return node
    }

    var resultStr:Array[String] = Array.empty

    def output_TreeNode(node:TreeNode):String = {
        set_resultStr(node, 0)
        return get_resultStr()
    }

    def set_resultStr(node:TreeNode, n:Int) {
        if (node == null)
            return
        
        if (resultStr.size <= n) {
            resultStr = resultStr :+ "(" + node.value.toString() + ")"
        } else
            resultStr(n) = resultStr(n) + ",(" + node.value.toString() + ")"

        if (node.left != null)
            set_resultStr(node.left, n + 1)
        if (node.right != null)
            set_resultStr(node.right, n + 1)

        return
    }

    def get_resultStr():String = {
        var outputStr:String = ""

        for (temp_str <- resultStr) {
            outputStr = outputStr + temp_str + "\n"
        }

        resultStr = Array.empty
        return outputStr
    }

    def tree2str(t:TreeNode):String = {
        if (t == null)
            return ""

        var resultStr:String = t.value.toString()

        if (t.left == null && t.right == null)
            return resultStr

        resultStr = resultStr + "(" + tree2str(t.left) + ")"
        if (t.right != null)
            resultStr = resultStr + "(" + tree2str(t.right) + ")"

        return resultStr
    }
}

File: Problems/0100_0199/0106_Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal/Project_Java/Operate_TreeNode.java
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class Operate_TreeNode {
    public TreeNode set_TreeNode(String[] flds) {
         return set_TreeNode(flds, 0, 0);
    }

    public TreeNode set_TreeNode(String[] flds, int depth, int pos) {
        if (flds.length == 0)
            return null;

        int cur_pos = 0;
        for (int i = 0; i < depth; ++i)
            cur_pos += (int)Math.pow(2, i);
        
        if (cur_pos + pos > flds.length - 1)
            return null;
        
        if (flds[cur_pos + pos].equals("null"))
            return null;

        try {
            TreeNode node = new TreeNode(Integer.parseInt(flds[cur_pos + pos]));
            node.left = set_TreeNode(flds, depth + 1, 2*pos);
            node.right = set_TreeNode(flds, depth + 1, 2*pos + 1);

            return node;
        } catch (Exception e) {
            System.out.println("\n" +  e + "\n" +
                              "set_TreeNode() Error ... flds[" + Integer.toString(cur_pos + pos) + "] = " + flds[cur_pos + pos] + "\n");
            System.exit(1);

            return null;
        }
    }

    List<String> resultStr;

    public String output_TreeNode(TreeNode node) {
        resultStr = new ArrayList<String>();

        String outStr = set_ResultStr(node, 0);
        resultStr.clear();

        return outStr;
    }

    private String set_ResultStr(TreeNode node, int n) {
        if (node == null)
            return "";
        
        if (resultStr.size() <= n)
            resultStr.add("(" + Integer.toString(node.val) + ")");
        else
            resultStr.set(n, resultStr.get(n) + ",(" + Integer.toString(node.val) + ")");

        if (node.left != null)
            set_ResultStr(node.left, n + 1);
        if (node.right != null)
            set_ResultStr(node.right, n + 1);

        String outputStr = "";
        for(String s : resultStr) {
            outputStr += s + "\n";
        }

        return outputStr;
    }

    public String Tree2str(TreeNode t) {
        if (t == null)
            return "";

        String resultStr = Integer.toString(t.val);

        if (t.left == null && t.right == null)
            return resultStr;

        resultStr += "(" + Tree2str(t.left) + ")";
        if (t.right != null)
            resultStr += "(" + Tree2str(t.right) + ")";

        return resultStr;
    }
}
